File: python3/200-number-of-islands.py
"""
intuition:
when a first "1" is found, use graph traversal to update the adjacent markers to "0" to indicate that island had been visited
visit each marker normally
"""

from typing import List
from collections import deque, defaultdict


def numIslands(grid: List[List[str]]) -> int:
    m = len(grid)
    n = len(grid[0])
    count = 0
    
    def dfs_visit(x, y):
        if x < 0 or x >= m or y < 0 or y >= n or grid[x][y] == "0":
            return
        grid[x][y] = "0"
        dfs_visit(x + 1, y)
        dfs_visit(x - 1, y)
        dfs_visit(x, y + 1)
        dfs_visit(x, y - 1)

    # def bfs_visit(x, y):
    #     grid[x][y] = "0"
    #     bfs_queue = deque()
    #     seen = set()

    #     bfs_queue.append((x,y))
    #     seen.add((x,y))

    #     while bfs_queue:
    #         node = bfs_queue.popleft()

    #         directions = [(-1,0), (1,0), (0,-1), (0,1)]
    #         for d_i, d_j in directions:
    #             new_i = node[0] + d_i
    #             new_j = node[1] + d_j

    #             if (0 <= new_i < m and 0 <= new_j < n):
    #                 # adjacent nodes
    #                 if ((new_i, new_j) not in seen) and (grid[new_i][new_j] == "1"):
    #                     grid[new_i][new_j] = "0"
    #                     seen.add((new_i, new_j))
    #                     bfs_queue.append((new_i, new_j))
    #                 pass

    for i in range(m):
        for j in range(n):
            if grid[i][j] == "1":
                dfs_visit(i, j)
                # bfs_visit(i, j)
                count += 1
    return count

grid = [
  ["1","1","0","0","0"],
  ["1","1","0","0","0"],
  ["0","0","1","0","0"],
  ["0","0","0","1","1"]
]

print(numIslands(grid))

File: cpp/200-number-of-islands.cpp
#include <print>
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

class Solution
{
public:
    int floodFill(int M, int N, vector<vector<char>> &grid, int i, int j)
    {
        queue<pair<int, int>> q;
        vector<pair<int, int>> directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
        q.push({i, j});

        while (!q.empty())
        {
            auto curr = q.front();
            q.pop();

            for (int i = 0; i < 4; i++)
            {
                int adj_i = curr.first + directions[i].first;
                int adj_j = curr.second + directions[i].second;

                if (0 > adj_i || adj_i >= M || 0 > adj_j || adj_j >= N || grid[adj_i][adj_j] == '0')
                {
                    continue;
                }
                grid[adj_i][adj_j] = '0';
                q.push({adj_i, adj_j});
            }
        }
        return 1;
    }
    int numIslands(vector<vector<char>> &grid)
    {
        int M = grid.size();
        int N = grid[0].size();
        // vector<vector<bool>> visited(M, vector<bool>(N, false));
        int count = 0;

        for (int i = 0; i < M; i++)
        {
            for (int j = 0; j < N; j++)
            {
                if (grid[i][j] == '1')
                {
                    count += floodFill(M, N, grid, i, j);
                }
            }
        }

        return count;
    }
};

int main()
{
    return 0;
}